lib/expr.ml

```diff
@@ -0,0 +1,280 @@
+    | Relop (op1, e1, e3), Relop (op2, e2, e4) ->
+      op1 = op2 && equal e1 e2 && equal e3 e4
+    | Triop (op1, e1, e3, e5), Triop (op2, e2, e4, e6) ->
+      op1 = op2 && equal e1 e2 && equal e3 e4 && equal e5 e6
+    | Symbol s1, Symbol s2 -> Symbol.equal s1 s2
+    | Extract (e1, h1, l1), Extract (e2, h2, l2) ->
+      equal e1 e2 && h1 = h2 && l1 = l2
+    | Concat (e1, e3), Concat (e2, e4) -> equal e1 e2 && equal e3 e4
+    | _ -> false
+
+let get_symbols (e : t list) : Symbol.t list =
+  let rec symbols e =
+    match e.e with
+    | Val _ -> []
+    | Ptr (_, offset) -> symbols offset
+    | Unop (_, e1) -> symbols e1
+    | Binop (_, e1, e2) -> symbols e1 @ symbols e2
+    | Triop (_, e1, e2, e3) -> symbols e1 @ symbols e2 @ symbols e3
+    | Relop (_, e1, e2) -> symbols e1 @ symbols e2
+    | Cvtop (_, e) -> symbols e
+    | Symbol s -> [ s ]
+    | Extract (e, _, _) -> symbols e
+    | Concat (e1, e2) -> symbols e1 @ symbols e2
+  in
+  List.fold_left
+    (fun accum x -> if List.mem x accum then accum else x :: accum)
+    []
+    (List.concat_map symbols e)
+
+let negate_relop ({ e; ty } : t) : (t, string) Result.t =
+  let e =
+    match e with
+    | Relop (Eq, e1, e2) -> Ok (Relop (Ne, e1, e2))
+    | Relop (Ne, e1, e2) -> Ok (Relop (Eq, e1, e2))
+    | Relop (Lt, e1, e2) -> Ok (Relop (Ge, e1, e2))
+    | Relop (LtU, e1, e2) -> Ok (Relop (GeU, e1, e2))
+    | Relop (Le, e1, e2) -> Ok (Relop (Gt, e1, e2))
+    | Relop (LeU, e1, e2) -> Ok (Relop (GtU, e1, e2))
+    | Relop (Gt, e1, e2) -> Ok (Relop (Le, e1, e2))
+    | Relop (GtU, e1, e2) -> Ok (Relop (LeU, e1, e2))
+    | Relop (Ge, e1, e2) -> Ok (Relop (Lt, e1, e2))
+    | Relop (GeU, e1, e2) -> Ok (Relop (LtU, e1, e2))
+    | _ -> Error "negate_relop: not a relop."
+  in
+  Result.map (fun relop -> { ty; e = relop }) e
+
+module Pp = struct
+  let rec pp fmt (e : t) =
+    let fprintf = Format.fprintf in
+    match e.e with
+    | Val v -> fprintf fmt "%a" Value.pp v
+    | Ptr (base, offset) -> fprintf fmt "(Ptr (i32 %ld) %a)" base pp offset
+    | Unop (op, e) -> fprintf fmt "(%a.%a %a)" Ty.pp e.ty pp_unop op pp e
+    | Binop (op, e1, e2) ->
+      fprintf fmt "(%a.%a %a %a)" Ty.pp e.ty pp_binop op pp e1 pp e2
+    | Triop (op, e1, e2, e3) ->
+      fprintf fmt "(%a.%a %a %a %a)" Ty.pp e.ty pp_triop op pp e1 pp e2 pp e3
+    | Relop (op, e1, e2) ->
+      fprintf fmt "(%a.%a %a %a)" Ty.pp e.ty pp_relop op pp e1 pp e2
+    | Cvtop (op, e) -> fprintf fmt "(%a.%a %a)" Ty.pp e.ty pp_cvtop op pp e
+    | Symbol s -> fprintf fmt "%a" Symbol.pp s
+    | Extract (e, h, l) -> fprintf fmt "(extract %a %d %d)" pp e l h
+    | Concat (e1, e2) -> fprintf fmt "(++ %a %a)" pp e1 pp e2
+
+  let pp_list fmt (es : t list) =
+    Format.pp_print_list ~pp_sep:Format.pp_print_space pp fmt es
+
+  let pp_smt fmt (es : t list) : unit =
+    let pp_symbols fmt syms =
+      Format.pp_print_list ~pp_sep:Format.pp_print_newline
+        (fun fmt sym ->
+          let x = Symbol.to_string sym in
```

Comment:
We don't have a `Symbol.pp` yet ?

Reply:
We do, I'll use it here instead :smiley: